api/v1alpha1/release_types_test.go

```diff
@@ -0,0 +1,446 @@
+				Type:   releaseConditionType,
+				Status: metav1.ConditionTrue,
+			}
+			Expect(r.HasSucceeded()).To(BeTrue())
+		})
+	})
+
+	Context("When IsDone method is called", func() {
+		It("should return false when condition struct is nil", func() {
+			Expect(r.IsDone()).To(BeFalse())
+		})
+
+		It("should return true when condition.Status is ConditionTrue", func() {
+			r.Status.Conditions[0] = metav1.Condition{
+				Type:   releaseConditionType,
+				Status: metav1.ConditionTrue,
+			}
+			Expect(r.IsDone()).To(BeTrue())
+		})
+
+		It("should return false when condition.Status is ConditionUnknown", func() {
+			r.Status.Conditions[0] = metav1.Condition{
+				Type:   releaseConditionType,
+				Status: metav1.ConditionUnknown,
+			}
+			Expect(r.IsDone()).To(BeFalse())
+		})
+	})
+
+	Context("When MarkFailed method is called", func() {
+		/*
+			|--------------------------------------------------------------------|
+			|                             NOOP matrix                            |
+			|--------------------------------------------------------------------|
+			| CompletionTime | condition.Status | time not nil | IsDone() | NOOP |
+			|----------------|------------------|--------------|----------|------|
+			|      zero      | ConditionUnknown |     true     |  false   |  no  |
+			|      zero      | ConditionTrue    |     true     |  true    |  yes |
+			|      nil       | ConditionUnknown |     false    |  false   |  no  |
+			|      nil       | ConditionTrue    |     false    |  true    |  no  |
+			|      Now()     | ConditionUnknown |     true     |  false   |  no  |
+			|      Now()     | ConditionTrue    |     true     |  true    |  yes |
+			|--------------------------------------------------------------------|
+
+			When no NOOP:
+			* expect zero or nil CompletionTime to be set to Now().
+			* expect setStatusConditionWithMessage() to append a new condition.
+			* * expect condition.Status: ConditionFalse
+			* * expect condition.Reason: to match the reason arg
+			* * expect condition.Message: to match the message arg
+		*/
+		It("should NOOP when IsDone() is true, and Status.CompletionTime is not nil", func() {
+			args := commonArgs{
+				reason:  ReleaseReasonValidationError,
+				message: "cow say m00",
+			}
+			r.Status.Conditions[0] = metav1.Condition{
+				Type:    releaseConditionType,
+				Status:  metav1.ConditionTrue,
+				Reason:  ReleaseReasonTargetDisabledError.String(),
+				Message: "Testcase one message string",
+			}
+			r.MarkFailed(args.reason, args.message)
+			Expect(len(r.Status.Conditions)).To(Equal(1))
+			Expect(r.Status.Conditions[0]).To(MatchFields(IgnoreMissing|IgnoreExtras, Fields{
+				"Status":  Equal(metav1.ConditionTrue),
+				"Type":    Equal(releaseConditionType),
+				"Reason":  Equal(ReleaseReasonTargetDisabledError.String()),
+				"Message": Equal("Testcase one message string"),
+			}))
+		})
+
+		It("should not NOOP when CompletionTime is nil, and IsDone returns false", func() {
```

Comment:
```suggestion
		It("should register a failed condition if the CompletionTime is nil", func() {
```
Also, we would need a similar test for when IsDone is true so we test both scenarios.